tput/tput.1
.Dd December 2008
.Dt TPUT 1

.Sh NAME

.Nm tput
.Nd change terminal characteristics

.Sh SYNOPSIS

.Nm tput
.Ar [-T type] operand...

.Sh DESCRIPTION

    The tput utility shall display terminal-dependent information. The manner
in which this information is retrieved is unspecified. The information
displayed shall clear the terminal screen, initialize the user's terminal, or
reset the user's terminal, depending on the operand given. The exact
consequences of displaying this information are unspecified.

.Sh OPTIONS

    The tput utility shall conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following option shall be supported:

    -T  type
        Indicate the type of terminal. If this option is not supplied and the
TERM variable is unset or null, an unspecified default terminal type shall be
used. The setting of type shall take precedence over the value in TERM .

.Sh OPERANDS

    The following strings shall be supported as operands by the
implementation in the POSIX locale:

    clear
        Display the clear-screen sequence.
    init
        Display the sequence that initializes the user's terminal in an
implementation-defined manner.
    reset
        Display the sequence that resets the user's terminal in an
implementation-defined manner.

    If a terminal does not support any of the operations described by these
operands, this shall not be considered an error condition.

.Sh STDIN

    Not used.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of tput:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]
    TERM
        Determine the terminal type. If this variable is unset or null, and
if the -T option is not specified, an unspecified default terminal type shall
be used.

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    If standard output is a terminal device, it may be used for writing the
appropriate sequence to clear the screen or reset or initialize the terminal.
If standard output is not a terminal device, undefined results occur.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        The requested string was written successfully.
     1
        Unspecified.
     2
        Usage error.
     3
        No information is available about the specified terminal type.
     4
        The specified operand is invalid.
    >4
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    If one of the operands is not available for the terminal, tput continues
processing the remaining operands.

The following sections are informative.
.Sh APPLICATION USAGE

    The difference between resetting and initializing a terminal is left
unspecified, as they vary greatly based on hardware types. In general,
resetting is a more severe action.

    Some terminals use control characters to perform the stated functions,
and on such terminals it might make sense to use tput to store the
initialization strings in a file or environment variable for later use.
However, because other terminals might rely on system calls to do this work,
the standard output cannot be used in a portable manner, such as the
following non-portable constructs:

    ClearVar=`tput clear`
    tput reset | mailx -s "Wake Up" ddg

.Sh EXAMPLES

        Initialize the terminal according to the type of terminal in the
environmental variable TERM . This command can be included in a .profile
file.

        tput init

        Reset a 450 terminal.

        tput -T 450 reset

.Sh RATIONALE

    The list of operands was reduced to a minimum for the following reasons:

        The only features chosen were those that were likely to be used by
human users interacting with a terminal.

        Specifying the full terminfo set was not considered desirable, but
the standard developers did not want to select among operands.

        This volume of IEEE Std 1003.1-2001 does not attempt to provide
applications with sophisticated terminal handling capabilities, as that falls
outside of its assigned scope and intersects with the responsibilities of
other standards bodies.

    The difference between resetting and initializing a terminal is left
unspecified as this varies greatly based on hardware types. In general,
resetting is a more severe action.

    The exit status of 1 is historically reserved for finding out if a
Boolean operand is not set. Although the operands were reduced to a minimum,
the exit status of 1 should still be reserved for the Boolean operands, for
those sites that wish to support them.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    stty, tabs



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



tput.c

/*
 * Copyright 2008-2009 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <curses.h>
#include <term.h>
#include <libpu.h>


static const char doc[] =
N_("tput - change terminal characteristics");

#define args_doc file_args_doc

static struct argp_option options[] = {
	{ "term", 'T', "type", 0,
	  N_("Indicate the type of terminal.") },
	{ }
};

static int tput_actor(struct walker *w, const char *fn, const struct stat *lst);
static int tput_pre_walk(struct walker *w);
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

static char *opt_term;

static struct walker walker;

static void tput_clear(void)
{
	tputs(clear_screen, lines > 0 ? lines : 1, putchar);
}

static const char *init_names[] = {
	"iprog",
	"is1",
	"is2",
	"if",
	"is3",
};

static const char *reset_names[] = {
	"rprog",
	"rs1",
	"rs2",
	"rf",
	"rs3",
};

static void tput_reset(bool is_init)
{
	const char **caps = is_init ? init_names : reset_names;
	char *val;

	val = xtigetstr(caps[0]);
	if (val)
		system(val);

	val = xtigetstr(caps[1]);
	if (val)
		write(STDOUT_FILENO, val, strlen(val));

	val = xtigetstr(caps[2]);
	if (val)
		write(STDOUT_FILENO, val, strlen(val));

	val = xtigetstr(caps[3]);
	if (val) {
		FILE *f;

		f = fopen(val, "r");
		if (f) {
			int ch;

			while ((ch = getc(f)) != EOF) {
				write(STDOUT_FILENO, &ch, 1);
			}

			fclose(f);
		}
	}

	val = xtigetstr(caps[4]);
	if (val)
		write(STDOUT_FILENO, val, strlen(val));

}

static int tput_actor(struct walker *w, const char *fn, const struct stat *lst)
{
	if (!strcmp(fn, "clear"))
		tput_clear();
	else if (!strcmp(fn, "init"))
		tput_reset(true);
	else if (!strcmp(fn, "reset"))
		tput_reset(false);

	return 0;
}

static int tput_pre_walk(struct walker *w)
{
	int err = 0;

	if (setupterm(opt_term, STDOUT_FILENO, &err) != OK) {
		fprintf(stderr, _("tput: setupterm failed\n"));
		return 1;
	}

	return 0;
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	PU_OPT_BEGIN

	case 'T':
		opt_term = arg;
		break;

	PU_OPT_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

int main (int argc, char *argv[])
{
	opt_term = get_terminal();
	walker.argp			= &argp;
	walker.cmdline_arg		= tput_actor;
	walker.pre_walk			= tput_pre_walk;
	return walk(&walker, argc, argv);
}

